src/device/common.h

```diff
@@ -226,22 +227,31 @@ static __forceinline__ __device__ void ncclRedopPtrDeref(struct ncclWorkElem* we
 }
 
 template<int SpecializedFnId, typename SpecializedRunWork, bool COLLTRACE>
-__forceinline__ __device__ void ncclKernelMain(struct ncclDevComm* comm, uint64_t channelMask, struct ncclWork* workHead) {
+__forceinline__ __device__ void ncclKernelMain(struct ncclDevComm* comm, struct channelMasks channelMask, struct ncclWork* workHead) {
   const int tid = threadIdx.x;
   int x = tid;
+  int total = 0, y;
+  int num = MAXCHANNELS/64 > 0 ? MAXCHANNELS/64 : 1;
+
   switch (tid/WARP_SIZE) {
   case 0:
-    if (channelMask & (1ull<<x)) {
-      int y = __popcll(channelMask & ((1ull<<x)-1));
+	  ncclShmem.channelId = blockIdx.x;
+  /*for (int i = 0; i < num; i++) {
+    if (channelMask.masks[i] & (1ull<<x)) {
+      y = __popcll(channelMask.masks[i] & ((1ull<<x)-1));
+      y = total + y;
       if (blockIdx.x == y) ncclShmem.channelId = x;
     }
-    if (WARP_SIZE < MAXCHANNELS) {
+    if (WARP_SIZE < 64) {
       x = WARP_SIZE + tid;
-      if (channelMask & (1ull<<x)) {
-        int y = __popcll(channelMask & ((1ull<<x)-1));
+      if (channelMask.masks[i] & (1ull<<x)) {
+	y = __popcll(channelMask.masks[i] & ((1ull<<x)-1));
+	y = y + total;
         if (blockIdx.x == y) ncclShmem.channelId = x;
       }
     }
+    total = __popcll(channelMask.masks[i]);
+  }*/
```

Comment:
channelId cannot be blockIdx for release version. It needs to be counted from channelMasks